--- 00errors-as-ok/main.go	2020-09-09 14:35:45.000000000 +0900
+++ 04errors-as-ok/main.go	2020-09-09 14:48:45.000000000 +0900
@@ -6,11 +6,15 @@
 	"os"
 )
 
+type hasTimeout interface {
+	Timeout() bool
+}
+
 func main() {
 	if _, err := os.Open("non-existing"); err != nil {
-		var pathError *os.PathError
+		var pathError hasTimeout
 		if errors.As(err, &pathError) {
-			fmt.Println("Failed at path:", pathError.Path)
+			fmt.Println("Failed at path:", pathError, "!!", pathError.Timeout())
 		} else {
 			fmt.Println(err)
 		}
