--- main00.go	2020-09-30 07:02:35.000000000 +0900
+++ main01.go	2020-09-30 07:11:17.000000000 +0900
@@ -6,6 +6,9 @@
 	"os"
 
 	_ "github.com/mattn/go-sqlite3"
+	"github.com/rs/zerolog"
+	sqldblogger "github.com/simukti/sqldb-logger"
+	"github.com/simukti/sqldb-logger/logadapter/zerologadapter"
 )
 
 func main() {
@@ -18,12 +21,24 @@
 		dbSource = "sample.db"
 	}
 
+	var db *sql.DB
 	db, err := sql.Open(driver, dbSource)
 	if err != nil {
 		panic(err.Error()) // Just for example purpose. You should use proper error handling instead of panic
 	}
 	defer db.Close()
 
+	logger := zerolog.New(
+		zerolog.ConsoleWriter{Out: os.Stdout, NoColor: false},
+	)
+	// populate log pre-fields here before set to
+	db = sqldblogger.OpenDriver(
+		dbSource,
+		db.Driver(),
+		zerologadapter.New(logger),
+		// optional config...
+	)
+
 	// Prepare statement for inserting data
 	stmtIns, err := db.Prepare("INSERT INTO squareNum VALUES( ?, ? )") // ? = placeholder
 	if err != nil {
