--- 00return/output	2019-10-02 22:32:24.371852131 +0900
+++ 01named/output	2019-10-02 22:32:24.598521064 +0900
@@ -53,10 +53,10 @@
 	0x0062 00098 (D/main.go:11)	MOVQ	16(SP), CX
 	0x0067 00103 (D/main.go:11)	PCDATA	$0, $2
 	0x0067 00103 (D/main.go:11)	XORPS	X0, X0
-	0x006a 00106 (D/main.go:11)	MOVUPS	X0, ""..autotmp_23+80(SP)
-	0x006f 00111 (D/main.go:11)	MOVQ	CX, ""..autotmp_23+80(SP)
+	0x006a 00106 (D/main.go:11)	MOVUPS	X0, ""..autotmp_21+80(SP)
+	0x006f 00111 (D/main.go:11)	MOVQ	CX, ""..autotmp_21+80(SP)
 	0x0074 00116 (D/main.go:11)	PCDATA	$2, $0
-	0x0074 00116 (D/main.go:11)	MOVQ	AX, ""..autotmp_23+88(SP)
+	0x0074 00116 (D/main.go:11)	MOVQ	AX, ""..autotmp_21+88(SP)
 	0x0079 00121 (D/main.go:11)	XCHGL	AX, AX
 	0x007a 00122 ($GOROOT/src/fmt/print.go:275)	PCDATA	$2, $1
 	0x007a 00122 ($GOROOT/src/fmt/print.go:275)	MOVQ	os.Stdout(SB), AX
@@ -68,7 +68,7 @@
 	0x008c 00140 ($GOROOT/src/fmt/print.go:275)	MOVQ	AX, 8(SP)
 	0x0091 00145 ($GOROOT/src/fmt/print.go:275)	PCDATA	$2, $1
 	0x0091 00145 ($GOROOT/src/fmt/print.go:275)	PCDATA	$0, $0
-	0x0091 00145 ($GOROOT/src/fmt/print.go:275)	LEAQ	""..autotmp_23+80(SP), AX
+	0x0091 00145 ($GOROOT/src/fmt/print.go:275)	LEAQ	""..autotmp_21+80(SP), AX
 	0x0096 00150 ($GOROOT/src/fmt/print.go:275)	PCDATA	$2, $0
 	0x0096 00150 ($GOROOT/src/fmt/print.go:275)	MOVQ	AX, 16(SP)
 	0x009b 00155 ($GOROOT/src/fmt/print.go:275)	MOVQ	$1, 24(SP)
@@ -94,10 +94,10 @@
 	0x00e3 00227 (D/main.go:12)	MOVQ	16(SP), CX
 	0x00e8 00232 (D/main.go:12)	PCDATA	$0, $3
 	0x00e8 00232 (D/main.go:12)	XORPS	X0, X0
-	0x00eb 00235 (D/main.go:12)	MOVUPS	X0, ""..autotmp_30+64(SP)
-	0x00f0 00240 (D/main.go:12)	MOVQ	CX, ""..autotmp_30+64(SP)
+	0x00eb 00235 (D/main.go:12)	MOVUPS	X0, ""..autotmp_26+64(SP)
+	0x00f0 00240 (D/main.go:12)	MOVQ	CX, ""..autotmp_26+64(SP)
 	0x00f5 00245 (D/main.go:12)	PCDATA	$2, $0
-	0x00f5 00245 (D/main.go:12)	MOVQ	AX, ""..autotmp_30+72(SP)
+	0x00f5 00245 (D/main.go:12)	MOVQ	AX, ""..autotmp_26+72(SP)
 	0x00fa 00250 (D/main.go:12)	XCHGL	AX, AX
 	0x00fb 00251 ($GOROOT/src/fmt/print.go:275)	PCDATA	$2, $1
 	0x00fb 00251 ($GOROOT/src/fmt/print.go:275)	MOVQ	os.Stdout(SB), AX
@@ -109,7 +109,7 @@
 	0x010d 00269 ($GOROOT/src/fmt/print.go:275)	MOVQ	AX, 8(SP)
 	0x0112 00274 ($GOROOT/src/fmt/print.go:275)	PCDATA	$2, $1
 	0x0112 00274 ($GOROOT/src/fmt/print.go:275)	PCDATA	$0, $0
-	0x0112 00274 ($GOROOT/src/fmt/print.go:275)	LEAQ	""..autotmp_30+64(SP), AX
+	0x0112 00274 ($GOROOT/src/fmt/print.go:275)	LEAQ	""..autotmp_26+64(SP), AX
 	0x0117 00279 ($GOROOT/src/fmt/print.go:275)	PCDATA	$2, $0
 	0x0117 00279 ($GOROOT/src/fmt/print.go:275)	MOVQ	AX, 16(SP)
 	0x011c 00284 ($GOROOT/src/fmt/print.go:275)	MOVQ	$1, 24(SP)
@@ -293,20 +293,20 @@
 	0x000f 00015 (D/main.go:20)	TESTQ	CX, CX
 	0x0012 00018 (D/main.go:20)	JNE	38
 	0x0014 00020 (D/main.go:21)	PCDATA	$0, $1
-	0x0014 00020 (D/main.go:21)	MOVQ	$0, "".~r2+32(SP)
+	0x0014 00020 (D/main.go:21)	MOVQ	$0, "".p+32(SP)
 	0x001d 00029 (D/main.go:21)	XORPS	X0, X0
-	0x0020 00032 (D/main.go:21)	MOVUPS	X0, "".~r2+40(SP)
+	0x0020 00032 (D/main.go:21)	MOVUPS	X0, "".p+40(SP)
 	0x0025 00037 (D/main.go:21)	RET
 	0x0026 00038 (D/main.go:23)	PCDATA	$2, $1
 	0x0026 00038 (D/main.go:23)	MOVQ	"".name+8(SP), DX
 	0x002b 00043 (D/main.go:23)	PCDATA	$2, $0
-	0x002b 00043 (D/main.go:23)	MOVQ	DX, "".~r2+32(SP)
-	0x0030 00048 (D/main.go:23)	MOVQ	AX, "".~r2+40(SP)
-	0x0035 00053 (D/main.go:23)	MOVQ	CX, "".~r2+48(SP)
+	0x002b 00043 (D/main.go:23)	MOVQ	DX, "".p+32(SP)
+	0x0030 00048 (D/main.go:23)	MOVQ	AX, "".p+40(SP)
+	0x0035 00053 (D/main.go:23)	MOVQ	CX, "".p+48(SP)
 	0x003a 00058 (D/main.go:23)	RET
-	0x003b 00059 (D/main.go:18)	MOVQ	$0, "".~r2+32(SP)
+	0x003b 00059 (D/main.go:18)	MOVQ	$0, "".p+32(SP)
 	0x0044 00068 (D/main.go:18)	XORPS	X0, X0
-	0x0047 00071 (D/main.go:18)	MOVUPS	X0, "".~r2+40(SP)
+	0x0047 00071 (D/main.go:18)	MOVUPS	X0, "".p+40(SP)
 	0x004c 00076 (D/main.go:18)	RET
 	0x0000 48 8b 44 24 10 48 85 c0 74 31 48 8b 4c 24 18 48  H.D$.H..t1H.L$.H
 	0x0010 85 c9 75 12 48 c7 44 24 20 00 00 00 00 0f 57 c0  ..u.H.D$ .....W.
@@ -363,12 +363,13 @@
 	rel 65+4 t=8 fmt.init+0
 	rel 71+4 t=15 "".initdone·+-1
 	rel 86+4 t=8 runtime.morestack_noctxt+0
-go.info."".create$abstract SDWARFINFO dupok size=33
+go.info."".create$abstract SDWARFINFO dupok size=41
 	0x0000 04 2e 63 72 65 61 74 65 00 01 01 11 6e 61 6d 65  ..create....name
 	0x0010 00 00 00 00 00 00 11 61 67 65 00 00 00 00 00 00  .......age......
-	0x0020 00                                               .
+	0x0020 11 70 00 01 00 00 00 00 00                       .p.......
 	rel 18+4 t=28 go.info.string+0
 	rel 28+4 t=28 go.info.int+0
+	rel 36+4 t=28 go.info."".Person+0
 go.info.fmt.Println$abstract SDWARFINFO dupok size=42
 	0x0000 04 66 6d 74 2e 50 72 69 6e 74 6c 6e 00 01 01 11  .fmt.Println....
 	0x0010 61 00 00 00 00 00 00 11 6e 00 01 00 00 00 00 11  a.......n.......
@@ -591,21 +592,26 @@
 	rel 44+4 t=6 type.*[1]interface {}+0
 	rel 48+8 t=1 type.interface {}+0
 	rel 56+8 t=1 type.[]interface {}+0
-go.loc."".create SDWARFLOC size=109
+go.loc."".create SDWARFLOC size=161
 	0x0000 ff ff ff ff ff ff ff ff 00 00 00 00 00 00 00 00  ................
 	0x0010 00 00 00 00 00 00 00 00 4d 00 00 00 00 00 00 00  ........M.......
 	0x0020 07 00 9c 93 08 91 08 93 08 00 00 00 00 00 00 00  ................
 	0x0030 00 00 00 00 00 00 00 00 00 ff ff ff ff ff ff ff  ................
 	0x0040 ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
 	0x0050 00 4d 00 00 00 00 00 00 00 02 00 91 10 00 00 00  .M..............
-	0x0060 00 00 00 00 00 00 00 00 00 00 00 00 00           .............
+	0x0060 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff  ................
+	0x0070 ff ff ff ff ff 00 00 00 00 00 00 00 00 1d 00 00  ................
+	0x0080 00 00 00 00 00 4d 00 00 00 00 00 00 00 02 00 91  .....M..........
+	0x0090 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
+	0x00a0 00                                               .
 	rel 8+8 t=1 "".create+0
 	rel 65+8 t=1 "".create+0
-go.info."".create SDWARFINFO size=54
+	rel 117+8 t=1 "".create+0
+go.info."".create SDWARFINFO size=51
 	0x0000 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
 	0x0010 00 00 00 00 00 01 9c 13 00 00 00 00 00 00 00 00  ................
-	0x0020 13 00 00 00 00 00 00 00 00 0f 7e 72 32 00 01 0f  ..........~r2...
-	0x0030 00 00 00 00 00 00                                ......
+	0x0020 13 00 00 00 00 00 00 00 00 13 00 00 00 00 00 00  ................
+	0x0030 00 00 00                                         ...
 	rel 1+4 t=28 go.info."".create$abstract+0
 	rel 5+8 t=1 "".create+0
 	rel 13+8 t=1 "".create+77
@@ -613,7 +619,8 @@
 	rel 28+4 t=28 go.loc."".create+0
 	rel 33+4 t=28 go.info."".create$abstract+22
 	rel 37+4 t=28 go.loc."".create+57
-	rel 48+4 t=28 go.info."".Person+0
+	rel 42+4 t=28 go.info."".create$abstract+32
+	rel 46+4 t=28 go.loc."".create+109
 go.range."".create SDWARFRANGE size=0
 go.isstmt."".create SDWARFMISC size=0
 	0x0000 04 05 01 05 02 05 01 05 02 09 01 09 02 05 01 10  ................
