--- output.full.py	2020-02-08 20:46:14.000000000 +0900
+++ output.simple.py	2020-02-08 20:46:14.000000000 +0900
@@ -1,13 +1,13 @@
-import typing as t
+
 
 def hello(*, name="world") -> str:
     return f"hello, {name}"
 
 
-def main(argv: t.Optional[t.List[str]] = None) -> t.Any:
+def main(argv=None):
     import argparse
 
-    parser = argparse.ArgumentParser(prog=hello.__name__, description=hello.__doc__, formatter_class=type('_HelpFormatter', (argparse.ArgumentDefaultsHelpFormatter, argparse.RawTextHelpFormatter), {}))
+    parser = argparse.ArgumentParser(prog=hello.__name__, description=hello.__doc__)
     parser.print_usage = parser.print_help  # type: ignore
     parser.add_argument('--name', required=False, default='world', help='-')
     args = parser.parse_args(argv)
