--- /tmp/before	2019-08-17 02:02:36.283775343 +0900
+++ /tmp/after	2019-08-17 01:59:47.203455490 +0900
@@ -19,7 +19,7 @@
     reraise(exc_type, exc_value, tb)
   File "/usr/local/lib/python3.8/site-packages/jinja2/_compat.py", line 37, in reraise
     raise value.with_traceback(tb)
-  File "00ng.j2", line 3, in template
+  File "00ng.j2", line 3, in <module>
     This is {{me|upper}.
   File "/usr/local/lib/python3.8/site-packages/jinja2/environment.py", line 497, in _parse
     return Parser(self, source, name, encode_filename(filename)).parse()
@@ -61,6 +61,6 @@
     self.current = next(self._iter)
   File "/usr/local/lib/python3.8/site-packages/jinja2/lexer.py", line 562, in wrap
     for lineno, token, value in stream:
-  File "/usr/local/lib/python3.8/site-packages/jinja2/lexer.py", line 683, in tokeniter
-    filename)
+  File "/usr/local/lib/python3.8/site-packages/jinja2/lexer.py", line 681, in tokeniter
+    raise TemplateSyntaxError('unexpected \'%s\'' %
 jinja2.exceptions.TemplateSyntaxError: unexpected '}'
